File: 2023/go/day5/day5.go
package day5
import (
    "strings"
"strconv"
"regexp"
"fmt"
)
// File.open("day5.txt").read.lines.map(&:strip)


// def parse(lines)
//   seeds = lines[0].split(' ').map(&:to_i)
//   seeds.shift
//   transforms = []
//   i = 1
//   loop do
//     break if i >= lines.size
//     l = lines[i]
//     i+=1
//     next if l.empty?


//     if l =~ /^\d+/
//       transforms[-1].push(l.split(' ').map(&:to_i))
//     else
//       transforms.push([])
//     end
//   end
//   [seeds, transforms]
// end

func split(data string) []string {
    rv := []string{}
    for _, line := range strings.Split(strings.TrimSuffix(data, "\n"), "\n") {
        rv = append(rv, line)
    }
    return rv
}

func parse(data string) ([]int, [][][]int){
    lines := split(data)
    seeds := []int{}
    for i, s := range strings.Split(lines[0], " ") {
        if i > 0 {
            i, _ := strconv.Atoi(s)
            seeds = append(seeds, i)
        }
    }

    transforms := [][][]int{}
    for idx, l := range lines{
        if idx == 0 || l == "" {
            continue
        }

        if match, _ := regexp.MatchString("^[0-9]", l); match {
            vals := []int{}
            for _, s := range strings.Split(l, " ") {
                v, _ :=  strconv.Atoi(s)
                vals = append(vals, v)
            }

            transforms[len(transforms) - 1 ] = append(transforms[len(transforms) - 1], vals)
        } else {
            transforms = append(transforms, [][]int{})
        }

    }
    return seeds, transforms
}

func Part1(data string) int {
    seeds, transforms := parse(data)
    min := -1
    for _, seed := range seeds {
        acc := seed
        for _, transform := range transforms {
            for _, t := range transform {
                dest := t[0]
                start := t[1]
                rng := t[2]
                if start <= acc && acc < start + rng {
                  acc = acc - start + dest  
                  break
                }
            }
        }
        if min < 0 || min > acc {
            min = acc
        }
    }
    return min
}

func Part2(data string) int {
    seed_ranges, transforms := parse(data)
    seeds := []int{}
    i := 0
    for {
        if i == len(seed_ranges) {
            break
        }
        a := seed_ranges[i]
        b := seed_ranges[i+1]
        q := 0
        i +=2
        for {
            if q == b {
                break
            }

            seeds = append(seeds, a + q)
            q += 1
        }

    }
    fmt.Printf("Trying out %d seeds", len(seeds))

    min := -1
    for _, seed := range seeds {
        acc := seed
        for _, transform := range transforms {
            for _, t := range transform {
                dest := t[0]
                start := t[1]
                rng := t[2]
                if start <= acc && acc < start + rng {
                  acc = acc - start + dest  
                  break
                }
            }
        }
        if min < 0 || min > acc {
            min = acc
        }
    }
    return min
}

// def part2(data)
//   seed_ranges, transforms = parse(data)
//   seeds = []
//   loop do
//     break if seed_ranges.empty?
//     print "."
//     s = seed_ranges.shift
//     r = seed_ranges.shift
//     seeds += (s..(s+r)).to_a
//   end

File: 2023/ruby/day5.rb
sample=<<~TXT.lines.map(&:strip)
seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
TXT

real = File.open("day5.txt").read.lines.map(&:strip)


def parse(lines)
  seeds = lines[0].split(' ').map(&:to_i)
  seeds.shift
  transforms = []
  i = 1
  loop do
    break if i >= lines.size
    l = lines[i]
    i+=1
    next if l.empty?


    if l =~ /^\d+/
      transforms[-1].push(l.split(' ').map(&:to_i))
    else
      transforms.push([])
    end
  end
  [seeds, transforms]
end

def part1(data)
  seeds, transforms = parse(data)
  seeds.map do |seed|
    # puts "For seed #{seed}"
    transforms.reduce(seed) do |acc, transform|
      transform.reduce(acc) do |acc2, (dest, start, range)|
        if start <= acc2 && acc2 < start + range
          v = acc2 - start + dest  
          # puts "Match on #{start}..#{start+range}: #{acc2 } => #{v}"
          break v
        else
          acc2
        end
      end
    end
  end.min
end

def part2(data)
  seed_ranges, transforms = parse(data)
  pairs = []
  loop do
    break if seed_ranges.empty?
    s = seed_ranges.shift
    r = seed_ranges.shift
    pairs.push([s, r])
  end
  
  input = pairs
  transforms.each do |transform|
    output = []
    transform.each do |(dest, start, range)|
      next_input = []
      input.each do |pair|
        # puts "Evaluating #{pair} against #{dest} / #{start} / #{range}"
        p_start, p_range = pair
        if start > p_start + p_range || start + range < p_start # start is after end of range, or p_start is after end of transform
          next_input.push(pair)
          # puts "not in range, no change"
          next
        end

        start_offset = 0
        if start > p_start
          next_input.push([p_start, start - p_start])
        else
          start_offset = p_start - start
        end

        interval_start = start + start_offset
        interval_end = [start + range, p_start + p_range].min

        output.push([interval_start - start + dest, interval_end - interval_start])

        if start + range < p_start + p_range
          next_input.push([interval_end, p_start + p_range - start - range])
        end

      end
      input = next_input
      # puts "next input is #{next_input}"
      # break
    end
    # puts "finished with input #{input} and output #{output}"
    # break;
    input += output
  end

  input.map(&:first).min

  # we start with a series of ranges and a set of transforms
  # for each transform, we need create a new series of ranges which serve as input to the next transform
  #   for each transform, we need to look at each range, and find the overlap with the transform range. We then transform the overlap
  # At the end, we can find the minimum value of any of the resulting ranges
end

puts part1(sample)
puts part1(real)

puts part2(sample)
puts part2(real)
